File: objectiveai-js/src/agent/completions/http.ts
import z from "zod";
import { ObjectiveAI, type RequestOptions } from "../../client";
import { Stream } from "../../stream";
import {
  AgentCompletionsRequestAgentCompletionCreateParamsSchema,
  type AgentCompletionsRequestAgentCompletionCreateParams,
} from "./request/agentCompletionCreateParams";
import type { AgentCompletionsRequestAgentCompletionNotifyParams } from "./request/agentCompletionNotifyParams";
import type { AgentCompletionsResponseUnaryAgentCompletion } from "./response/unary/agentCompletion";
import type { AgentCompletionsResponseStreamingAgentCompletionChunk } from "./response/streaming/agentCompletionChunk";

export const AgentCompletionsRequestAgentCompletionCreateParamsStreamingSchema =
  AgentCompletionsRequestAgentCompletionCreateParamsSchema.extend({
    stream: z.literal(true),
  });
export type AgentCompletionsRequestAgentCompletionCreateParamsStreaming = z.infer<
  typeof AgentCompletionsRequestAgentCompletionCreateParamsStreamingSchema
>;

export const AgentCompletionsRequestAgentCompletionCreateParamsUnarySchema =
  AgentCompletionsRequestAgentCompletionCreateParamsSchema.extend({
    stream: z.literal(false).optional().nullable(),
  });
export type AgentCompletionsRequestAgentCompletionCreateParamsUnary = z.infer<
  typeof AgentCompletionsRequestAgentCompletionCreateParamsUnarySchema
>;

export function agentCompletionsCreateAgentCompletion(
  client: ObjectiveAI,
  body: AgentCompletionsRequestAgentCompletionCreateParamsStreaming,
  options?: RequestOptions,
): Promise<Stream<AgentCompletionsResponseStreamingAgentCompletionChunk>>;
export function agentCompletionsCreateAgentCompletion(
  client: ObjectiveAI,
  body: AgentCompletionsRequestAgentCompletionCreateParamsUnary,
  options?: RequestOptions,
): Promise<AgentCompletionsResponseUnaryAgentCompletion>;
export function agentCompletionsCreateAgentCompletion(
  client: ObjectiveAI,
  body: AgentCompletionsRequestAgentCompletionCreateParams,
  options?: RequestOptions,
): Promise<
  | Stream<AgentCompletionsResponseStreamingAgentCompletionChunk>
  | AgentCompletionsResponseUnaryAgentCompletion
> {
  if (body.stream) {
    return client.post_streaming<AgentCompletionsResponseStreamingAgentCompletionChunk>(
      "/agent/completions",
      body,
      options,
    );
  }
  return client.post_unary<AgentCompletionsResponseUnaryAgentCompletion>(
    "/agent/completions",
    body,
    options,
  );
}

/**
 * Notify a running agent completion with a user message.
 *
 * Pushes a `RichContent` payload at the agent completion identified
 * by `params.response_id`; the api queues it and surfaces it to the
 * model on its next natural inspection point. There is no response
 * body — any 2xx status is the success signal.
 */
export function agentCompletionsNotifyAgentCompletion(
  client: ObjectiveAI,
  body: AgentCompletionsRequestAgentCompletionNotifyParams,
  options?: RequestOptions,
): Promise<void> {
  return client.post_unary_no_response(
    "/agent/completions/notify",
    body,
    options,
  );
}

File: objectiveai-sdk-py/objectiveai_sdk/agent/completions/http.py
"""HTTP functions for agent completions."""

from __future__ import annotations

from typing import TYPE_CHECKING, Union

if TYPE_CHECKING:
    from objectiveai_sdk.agent.completions.request import AgentCompletionCreateParams
    from objectiveai_sdk.agent.completions.request.agent_completion_notify_params import (
        AgentCompletionNotifyParams,
    )
    from objectiveai_sdk.agent.completions.response.streaming import AgentCompletionChunk
    from objectiveai_sdk.agent.completions.response.unary import AgentCompletion
    from objectiveai_sdk.client import ObjectiveAI
    from objectiveai_sdk.stream import Stream


async def create_agent_completion(
    client: ObjectiveAI,
    params: AgentCompletionCreateParams,
) -> Union[AgentCompletion, Stream[AgentCompletionChunk]]:
    """Create an agent completion.

    If ``params.stream`` is true, returns a streaming response.
    Otherwise returns the complete response.
    """
    if getattr(params, "stream", None):
        return await client.post_streaming("agent/completions", params)
    return await client.post_unary("agent/completions", params)


async def notify_agent_completion(
    client: ObjectiveAI,
    params: AgentCompletionNotifyParams,
) -> None:
    """Notify a running agent completion with a user message.

    Pushes a ``RichContent`` payload at the agent completion identified
    by ``params.response_id``; the api queues it and surfaces it to
    the model on its next natural inspection point. There is no
    response body — any 2xx status is the success signal.
    """
    await client.post_unary_no_response("agent/completions/notify", params)
